File: 2015/day14.py
# keywords: hash/dict, array

import numpy as np

racers = {}

for j in open("Day14_input.txt"):
  parts = j.strip().split(' ')
  who = parts[0]
  km = int(parts[3])
  flyfor = int(parts[6])
  rest = int(parts[13])
  racers[who] = { 'km':km, 'flyfor':flyfor, 'rest':rest, 'next':1, 'score':0, 'dist':0 }

# for racer in racers:
#   print(racer,":",racers[racer])

def getNext(sec, r):
  if sec <= r['flyfor']:
    x = y = sec % r['flyfor']
  else:
    x = int(sec / (r['flyfor'] + r['rest']))
    y = (sec - ((r['flyfor'] + r['rest']) * x)) % r['flyfor']
  if y == 0:
    return r['rest'] + 1
  else:
    return 1

sec = 1
while sec < 20:

  for r in racers:
    racer = racers[r]
    if sec == racer['next']:
      racer['dist'] += racer['km']    
      racer['next'] += getNext(sec, racer)

  d = []
  for r in racers: 
    d.append(racers[r]['dist'])
  print('d:', d)

  for v in racers.values():
    print("values:", v)
  # input()
  sec += 1
  exit()

print(racers['Dancer'])

File: 2015/Day14.rb
# keywords: hash, max_by, select

race = 2503
h = {}

File.readlines('Day14_input.txt').each do |line|
  parts = line.split(' ')
  
  a = parts[3].to_i  #flying distance
  b = parts[6].to_i  #travel time
  c = parts[13].to_i #rest
  d = parts[0]       #reindeer
  # puts "Reindeer: #{d}, flies at #{a} for #{b} with #{c} rest"
  
  h ||= []
  h[d] = { 'km': a, 'for': b , 'rest': c, 'next': 1, 'score': 0, 'dist': 0 }
end

# pp h

# Part 1
# h.each do |k, v|
#   cc = distance = 0
#   while cc + v[:b] < 2503 do
#     cc += v[:b]
#     distance += v[:a] * v[:b]
#     cc += v[:c]
#   end

  # x = (race / (b + c)).to_i
  # y = (race - ((b + c) * x))
  # z = if y > b
  #       1
  #     else
  #       0
  #     end

  # h ||= []
  # h[d] = (x + z) * (a*b)
  # puts "Distance travelled: #{h[d]}
  # "
# end

def getNext(sec, v)
  if sec <= v[:for]
    x = y = sec % v[:for]
  else
    x = (sec / (v[:for] + v[:rest])).to_i
    y = (sec - ((v[:for] + v[:rest]) * x)) % v[:for]
  end
  
  if y == 0
    v[:rest] + 1
  else
    1
  end
end

# Part 2
(1..2503).each do |sec|
  h.each do |k, v|
    # puts "Sec: #{sec}"
    if v[:next] == sec
      v[:dist] += v[:km]
      v[:next] += getNext(sec, v)
    end
  end
  max_dist = h.max_by { |k, v| v[:dist] }[1][:dist]
  leaders = h.select { |k, v| v[:dist] == max_dist }.map { |d| d[0] }
  leaders.each { |l| h[l][:score] += 1 }
  # pp "Leaders: #{leaders}"
  if sec % 200 == 0
    puts "Sec: #{sec}"
    h.each { |k, v| puts "#{k}: #{v}" }
    readline()
  end
end

pp h